### ISSUE TITLE
Cannot identify a component in rule snippet

### ISSUE THREAD

**Issue by GCheung55** [embroider-build/embroider#471]
Embroider is having issues with components from `ember-power-select`.

https://github.com/cibernox/ember-power-select/issues/1368

> cannot identify a component in rule snippet: "{{power-select/before-options}}"

I believe the issue is that Resolver's `standardDasherize` method is not checking if the `ast.type` is a "Template". https://github.com/embroider-build/embroider/blob/master/packages/compat/src/resolver.ts#L224

The same error is also appearing for ember-basic-dropdown.

> cannot identify a component in rule snippet: "{{basic-dropdown-trigger}}"

**Comment by GCheung55** [comment]
@ef4 I'm having trouble writing a test to reproduce the issue. It looks like the tests for `resolver` aren't flexible to allow defining the templates of multiple components, which I think is necessary to reproduce the issue.

What I'm trying to do is mimic what `power-select` and `basic-dropdown` are doing, where they have a component that yields a hash that contains components based on a condition like

```
{{yield Component=(component (or @triggerComponent "basic-dropdown-trigger") ) }}
```

What I've tried doing so far:

```
  test.only('optional component missing in yield', function() {
    let findDependencies = configure({
      staticComponents: true,
      staticHelpers: true,
      packageRules: [
        {
          package: 'the-test-package',
          addonModules: {
            'components/this-one.js': {
              dependsOnComponents: ['{{that-one}}', '{{other-one}}'],
            },
          },
          components: {
            '{{this-one}}': { safeToIgnore: false },
          },
        },
      ],
    });
    givenFile('helpers/or.js');
    givenFile('components/this-one.js');
    let received = findDependencies(
      'templates/application.hbs',
      `{{yield (hash C=(component (or this.someOther "this-one") ) ) }}`
    );

    console.log('received', received);
    expect(received).toEqual([]);
  });
```

The problem with this ^ is that it's not triggering the resolver's `standardDasherize` method so I can't validate that setting `ast.type` is "Template" will address the issue.

**Comment by GCheung55** [comment]
Maybe I was looking at the wrong place since `dependency-rules` triggers the `resolver.resolveComponentSnippet` method.

This screenshot was shared in Discord: 
![image](https://user-images.githubusercontent.com/46033/84358144-9063ad00-ab7b-11ea-9eee-dcf3f01f11d9.png)

The snippet is `{{basic-dropdown-trigger}}`. `ember-basic-dropdown` doesn't explicitly use this in any of the templates, which leads me to think this is something in `ember-basic-dropdown` dependency rules.

**Comment by nightire** [comment]
@GCheung55  Doesn't this line use `basic-dropdown-trigger` explicitly?

https://github.com/cibernox/ember-basic-dropdown/blob/6a7fd3b102c5381f5c931c25f928e9924a64d4d5/addon/templates/components/basic-dropdown.hbs#L8

**Comment by nightire** [comment]
BTW, It was me shared the above screenshot in Discord, I should commit an issue with a minimal reproduction in last weekend, but I was busy ATM until now. Here's what I did for a temporary workaround:

```typescript
- if (ast.type === 'Program' && ast.body.length > 0) {
+ if ((ast.type === 'Program' || ast.type === 'Template') && ast.body.length > 0)
```

**Comment by GCheung55** [comment]
> @GCheung55 Doesn't this line use `basic-dropdown-trigger` explicitly?
> 
> https://github.com/cibernox/ember-basic-dropdown/blob/6a7fd3b102c5381f5c931c25f928e9924a64d4d5/addon/templates/components/basic-dropdown.hbs#L8

 I do not think so because the first parameter passed to the `component` helper is the output of the `or` helper. My guess is that due to this, the parsing done by the resolver doesn't "see"  `basic-dropdown-trigger` as a component dependency.

I've tried in some of the resolver test to reproduce the behavior with a template that looks like `{{yield C=(component (or this.otherComponent "this-one"))}}` but the result was that `this-one` was not found as a component dependency.

**Comment by GCheung55** [comment]
> BTW, It was me shared the above screenshot in Discord, I should commit an issue with a minimal reproduction in last weekend, but I was busy ATM until now. Here's what I did for a temporary workaround:

I wasn't sure if you had the same user name in Github, so I hesitated to tag you. :)

> ```js
> - if (ast.type === 'Program' && ast.body.length > 0) {
> + if ((ast.type === 'Program' || ast.type === 'Template') && ast.body.length > 0)
> ```

That's what I'm adding, just need to get a test to go through with "Template" as the AST type.

**Comment by GCheung55** [comment]
I've narrowed down the issue and discovered that there was a change to `ember-source/vendor/ember/ember-template-compiler` in `ember-source@v3.17.0` that is causing the issue. 

What I did to reproduce the issue was to change the `ember-source` version to 3.17.0 in `test-packages/support/package.json` and run the tests.

So it seems I don't need to write a test if the build fails with 3.17.0 and later.

I've created PR #472 to reproduce the issue. @ef4 how are the dependencies in `test-packages/support` updated? Is it enough to manually update `ember-source` to the latest? At the time of this comment, it's ~3.18.0.
